cmaes/_ipop.py

```diff
@@ -0,0 +1,164 @@
+        self._tolx = 1e-11
+
+        # termination criterion: tolerance in function value [1e-17, ~1e-3]
+        # pycma's default is 1e-11 though paper recommends 1e-12.
+        self._tolfun = 1e-11
+        self._funcval_generation_diff = 10 + math.ceil(30 * len(mean) / popsize)
+        # stores 'best' and 'worst'
+        self._funcval_values = np.empty(self._funcval_generation_diff * 2)
+
+    @property
+    def dim(self) -> int:
+        """A number of dimensions"""
+        return self._cma.dim
+
+    @property
+    def population_size(self) -> int:
+        """A population size"""
+        return self._popsize
+
+    @property
+    def generation(self) -> int:
+        """Generation number which is monotonically incremented
+        when multi-variate gaussian distribution is updated."""
+        return self._cma.generation
+
+    def ask(self) -> np.ndarray:
+        """Sample a parameter."""
+        return self._cma.ask()
+
+    def tell(self, solutions: List[Tuple[np.ndarray, float]]) -> None:
+        """Tell evaluation values."""
+        self._cma.tell(solutions)
+
+        self._update_objective_values([s[1] for s in solutions])
+
+        if self._should_stop():
+            self._n_restarts += 1
+            self._popsize *= self._inc_popsize
+            self._cma = CMA(popsize=self._popsize, **self._cma_opts)
+
+    def _update_objective_values(self, values: List[float]) -> None:
+        # Stores best values and worst values
+        # of the last '10 + math.ceil(30 * n / popsize)' generations.
+        np_values = np.array(values, dtype=float)
+        best_evaluation = np.min(np_values)
+        worst_evaluation = np.max(np_values)
+        value_index = 2 * (self._cma.generation % self._funcval_generation_diff)
+        self._funcval_values[value_index] = best_evaluation
+        self._funcval_values[value_index + 1] = worst_evaluation
+
+    def _should_stop(self) -> bool:
+        # eigen-decomposition
+        C = (self._cma._C + self._cma._C.T) / 2
+        D2, B = np.linalg.eigh(C)
+        D = np.sqrt(D2)
+        dC = np.diag(C)
+
+        # Update cache of B and D
+        self._cma._B, self._cma._D = B, D
+
+        # aliases
+        dim = self._cma.dim
+        mean = self._cma._mean
+        sigma = self._cma._sigma
+        pc = self._cma._pc
+
+        # Equal objective function values.
+        if np.max(self._funcval_values) - np.min(self._funcval_values) < self._tolfun:
+            return True
+
+        # TolX:
+        if np.all(sigma * dC < self._tolx) and np.all(sigma * pc < self._tolx):
+            return True
+
+        # Check axis and coordinates one by one on each generation.
+        i = self.generation % self._cma.dim
+
+        # No effect axis
+        if np.sum(mean - (0.1 * sigma * D[i] * B[:, i])) == dim:
+            return True
+
+        # No effect coordinates
+        if any(np.nonzero(mean == mean + (0.2 * sigma * np.sqrt(dC)))):
+            return True
```

Comment:
These lines are corresponds with the following on pycma:

https://github.com/CMA-ES/pycma/blob/af8bd709c7982c288e364c99ddb22ad3c2946f0b/cma/evolution_strategy.py#L3483-L3501